### DISCUSSION TITLE
Save FTS Index to Disk?

### DISCUSSION THREAD

**Discussion by thatcort** [duckdb/duckdb#11479] [General]
Hi, is it possible to create a full-text search index and save it to be used later? We'd like to be able to generate the indices in advance and have them loaded when needed. In particular we're looking at storing to a Delta table in object storage for long term persistence and Arrow IPC for short term persistence.

Thank you!

**Comment by Tishj** [comment]
https://duckdb.org/docs/extensions/full_text_search

**Comment by arjenpdevries** [comment]
If the database is on-disk, the FTS index is too, I believe. Did you start duckdb with a database file as parameter?

The index is just a few tables, so you can import / export as you like.

**Comment by thatcort** [comment]
Thanks @arjenpdevries. We are storing our data in delta tables on object storage and then importing the data into Arrow format.

Suppose we save the FTS index tables to disk. When we later read them, do we need to do any configuration for them to be used during FTS search, or is it sufficient that they exist in memory with the same naming convention?

I guess we can figure out the table names from reading the source code. Will this naming convention be stable between releases?

**Reply by arjenpdevries** [reply]
The naming convention is sufficient, the retrieval code is really a macro referring those tables. (The created tables are part of a schema whose name relates to the index.)

Consistency between releases I do not know, but expect no problems; maybe @lnkuiper can comment.

**Reply by lnkuiper** [reply]
The FTS index is just DuckDB tables stored in a schema with a retrieval macro, as Arjen says. This means that they will be persisted in storage (not just exist in memory). Once you create the FTS index, the original input tables are no longer needed:
```sql
create table test(id int, txt varchar);
insert into test values (42, 'my cool text');
pragma create_fts_index('test', 'id', 'txt');
drop table test;
select fts_main_test.match_bm25(1, 'hi');
-- macro still works after dropping 'test' returns NULL
create table test(id int, txt varchar);
insert into test values (42, 'my cool text');
-- macro can be used properly again after re-creating the table
select score, id, txt from (select *, fts_main_test.match_bm25(id, 'cool') as score from test);
-- ┌──────────────────────┬───────┬──────────────┐
-- │        score         │  id   │     txt      │
-- │        double        │ int32 │   varchar    │
-- ├──────────────────────┼───────┼──────────────┤
-- │ -0.47712125471966244 │    42 │ my cool text │
-- └──────────────────────┴───────┴──────────────┘
```
The FTS extension and its table names have been stable between releases for a long time now. I cannot guarantee that they will never change, but we do not have plans to change them as of now.

**Reply by thatcort** [reply]
@lnkuiper , @arjenpdevries, thank you! This is very helpful!
